### PR TITLE
SidePanel Submodules: Status for top module

### PR DESCRIPTION
**Pull request by gerhardol**
## Proposed changes 

Show status for top module in the side panel.
If there are changes in any submodules (commit or dirty) or changes in the topmodule itself, this status should be seen in the top most icon.
(The top module cannot be ahead/behind only dirty but will be so for any changes)

Minor refactorings and removal of unused code too, to be squashed.

### Limitations
There are some limitations to the implementation, due to performance issues (git-status can require some resources).

git-status reuses the call for the commit button and artificial commits in the grid, so it is executed for the current module only. If there are changes, the side panel status is currently refreshed only for the current submodule (and below).
At the initial opening, the status is calculated for all modules though, except files in the current top module.
There is also a limiter to not refresh the status too often (ignore updates within 15s).

If the top module is current, the git-status updates are sufficient to set the status for the top module sufficient. (When rewriting this section about the limitations, I updated the implementation too.)

If another submodule is current, git-status cannot be used to refresh all status. (Previously no update at all was done.) If a submodule has changes, also the top module is dirty. However, it is not possible to determine if the status for the top module (and all modules in between) can be cleared as the modules and files in the top module is not calculated. The status is therefore is therefore additive, i.e. 'dirty' is set but not cleared.

Summary limitations if top module is not current:
 * If only files in top module is changed, top is not set as dirty (not even initially)
 * Status for top module is changed to dirty, not cleared if there are changes
 * Changes in modules outside the tree is not monitored.
 * If the current submodule is level 3 or lower, the upper modules are not changed to dirty if current module is changed (this could be changed though).

### Before

image

### After

image

## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/Submodules/SubmoduleStatusProvider.cs [hunk 1]
```diff
@@ -326,6 +338,24 @@ private async Task UpdateSubmodulesStatusAsync(GitModule module, [CanBeNull] IRe
             }
         }
 
+        /// <summary>
+        /// Set the top module as dirty
+        /// If any module is changed
+        /// </summary>
+        /// <param name="module">current Git module</param>
+        private void SetTopModuleAsDirty(GitModule module)
```
Comment (RussKie) [inline GitCommands/Submodules/SubmoduleStatusProvider.cs]:
```suggestion
        private void SetTopModuleAsDirty(string submoduleWorkingDir)
```

File: GitCommands/Submodules/SubmoduleStatusProvider.cs [hunk 2]
```diff
@@ -381,6 +411,11 @@ private async Task GetSubmoduleDetailedStatusAsync(GitModule superModule, string
                     AddedAndRemovedText = submoduleStatus.AddedAndRemovedString()
                 };
 
+            if (submoduleStatus != null)
+            {
+                SetTopModuleAsDirty(superModule);
```
Comment (RussKie) [inline GitCommands/Submodules/SubmoduleStatusProvider.cs]:
```suggestion
                SetTopModuleAsDirty(superModule.GetTopModule().WorkingDir);
```

File: GitCommands/Submodules/SubmoduleStatusProvider.cs [hunk 3]
```diff
@@ -312,6 +317,13 @@ private async Task UpdateSubmodulesStatusAsync(GitModule module, [CanBeNull] IRe
             _previousSubmoduleUpdateTime = DateTime.Now;
             await TaskScheduler.Default;
 
+            // TopModule is dirty if there are any changes in any module
+            if (gitStatus != null
+                && gitStatus.Count > 0)
+            {
+                SetTopModuleAsDirty(module);
```
Comment (RussKie) [inline GitCommands/Submodules/SubmoduleStatusProvider.cs]:
```suggestion
                SetTopModuleAsDirty(module.GetTopModule().WorkingDir);
```

File: GitCommands/Submodules/SubmoduleStatusProvider.cs (hunk 1 above)
Comment (gerhardol) [reply GitCommands/Submodules/SubmoduleStatusProvider.cs]: changing

**Comment by gerhardol** [comment]
Force pushed RussKies suggestion, updating description

File: GitCommands/Submodules/SubmoduleStatusProvider.cs [hunk 4]
```diff
@@ -326,6 +338,24 @@ private async Task UpdateSubmodulesStatusAsync(GitModule module, [CanBeNull] IRe
...
         }
 
+        /// <summary>
+        /// Set the top module as dirty
+        /// If any module is changed
+        /// </summary>
+        /// <param name="topModuleWorkingDir">path to top module</param>
+        private void SetTopModuleAsDirty(string topModuleWorkingDir)
+        {
+            if (_submoduleInfos[topModuleWorkingDir].Detailed == null)
```
Comment (RussKie) [inline GitCommands/Submodules/SubmoduleStatusProvider.cs]: I missed this on the first review. Are you sure `_submoduleInfos[topModuleWorkingDir]` is safe?

Comment (gerhardol) [reply GitCommands/Submodules/SubmoduleStatusProvider.cs]: Yes, added on line 111

**Comment by gerhardol** [comment]
When updating the description about limitations, I realized that one of them could be eliminated so I updated slightly.

**Comment by gerhardol** [comment]
@msftbot merge in 2 days

**Comment by gerhardol** [comment]
@msftbot merge in 20 hours

**Pull request merged**